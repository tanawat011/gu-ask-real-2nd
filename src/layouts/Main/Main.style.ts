import type { SidebarState } from 'recoils/atoms'

import tw, { styled } from 'twin.macro'

import { DARK_THEME, FLEX, LIGHT_THEME } from 'constants/twTheme'

export const MainContainer = styled.div(({ fullSidebar }: SidebarState) => [
  [DARK_THEME.BG.COLOR, DARK_THEME.TEXT.COLOR],
  [LIGHT_THEME.BG.COLOR, LIGHT_THEME.TEXT.COLOR],
  [FLEX.ROW],
  tw`relative`,
  fullSidebar ? tw`h-screen` : tw`min-h-screen`,
])
export const MainContent = styled.div(() => [
  [DARK_THEME.BG.CONTENT_COLOR],
  [LIGHT_THEME.BG.CONTENT_COLOR],
  [FLEX.COL],
  tw`w-full`,
])
export const MainContentContainer = styled.div(() => [
  [FLEX.COL],
  tw`relative h-full overflow-x-auto`,
])
