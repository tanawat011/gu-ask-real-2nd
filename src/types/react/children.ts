export type Children = { children: JSX.Element }
