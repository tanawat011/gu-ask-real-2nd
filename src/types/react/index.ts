export * from './children'
export * from './reactRouterDom'
