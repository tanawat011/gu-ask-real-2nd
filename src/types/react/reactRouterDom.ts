import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
