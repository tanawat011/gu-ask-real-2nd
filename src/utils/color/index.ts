export * from './hex2Rgb'
export * from './rgb2Hex'
