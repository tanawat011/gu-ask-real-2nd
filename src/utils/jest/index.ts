export * from './renderWithProviders'
export * from './twColor'
