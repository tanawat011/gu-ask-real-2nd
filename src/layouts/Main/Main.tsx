import type { OutletContextProps } from 'types'

import type { UIEvent } from 'react'
import { useState, useEffect } from 'react'

import { Outlet, useLocation, useNavigate } from 'react-router-dom'
import { useRecoilState } from 'recoil'

import { Navbar } from 'components/Navbar'
import { Sidebar } from 'components/Sidebar'
import { useRouteList } from 'hooks/useRouteList'
import { useSidebarMenuList } from 'hooks/useSidebarMenuList'
import { sidebarAtom } from 'recoils/atoms/sidebar'

import { MainContainer, MainContent, MainContentContainer } from './Main.style'

export const MainLayout: React.FC = () => {
  const location = useLocation()
  const navigate = useNavigate()

  const [scroll, setScroll] = useState({
    top: 0,
    left: 0,
    height: 0,
    width: 0,
  })

  const [{ fullSidebar }, setFullSidebar] = useRecoilState(sidebarAtom)

  const [menuList, setMenuList] = useSidebarMenuList()
  const { main } = useRouteList()

  const { home } = main.children

  useEffect(() => {
    handleRedirectToHome()
  }, [])

  const handleRedirectToHome = () => {
    const defaultPaths = ['', '/']

    if (defaultPaths.includes(location.pathname)) {
      navigate(home.fullPath)
    }
  }

  const handleToggleSidebar = () => {
    setFullSidebar({ fullSidebar: !fullSidebar })
  }

  const handleScroll = (event: UIEvent<HTMLElement>) => {
    const { scrollTop, scrollLeft, scrollHeight, scrollWidth } =
      event.currentTarget

    setScroll({
      top: scrollTop,
      left: scrollLeft,
      height: scrollHeight,
      width: scrollWidth,
    })
  }

  return (
    <MainContainer fullSidebar={fullSidebar}>
      <Sidebar
        menuList={menuList}
        setMenuList={setMenuList}
        fullSidebar={fullSidebar}
      />

      <MainContent>
        <Navbar handleToggleSidebar={handleToggleSidebar} />

        <MainContentContainer
          data-testid='scrollable-content'
          id='main-layout'
          onScroll={handleScroll}
        >
          <Outlet context={{ scroll } as OutletContextProps} />
          <div className='min-h-[64px]'>Container Footer Template</div>
        </MainContentContainer>
      </MainContent>
    </MainContainer>
  )
}
