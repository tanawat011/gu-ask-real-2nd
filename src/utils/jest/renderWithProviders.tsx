import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
