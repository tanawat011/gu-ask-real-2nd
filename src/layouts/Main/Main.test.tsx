import { fireEvent, screen } from '@testing-library/react'

import { renderWithProviders } from 'utils/jest'

import { MainLayout } from './Main'

jest.mock('components/Sidebar', () => ({
  Sidebar: ({ setMenuList }: { setMenuList: () => void }) => {
    return (
      <div>
        <button data-testid='btn2' onClick={setMenuList}>
          Menu Item
        </button>
      </div>
    )
  },
}))

jest.mock('components/Navbar', () => ({
  Navbar: ({ handleToggleSidebar }: { handleToggleSidebar: () => void }) => {
    return (
      <div>
        <button data-testid='btn1' onClick={handleToggleSidebar}>
          Toggle Sidebar
        </button>
      </div>
    )
  },
}))

describe('<MainLayout />', () => {
  test('should render to match snapshot', () => {
    expect.assertions(1)

    const { asFragment } = renderWithProviders(<MainLayout />)

    const toggleSidebarButton = screen.getByTestId(/btn1/i)

    fireEvent.click(toggleSidebarButton)

    const menuItemButton = screen.getByTestId(/btn2/i)

    fireEvent.click(menuItemButton)

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render to wrong location path', () => {
    expect.assertions(1)

    window.location.assign('http://localhost:3000/xxx')

    const { asFragment } = renderWithProviders(<MainLayout />)

    expect(asFragment()).toMatchSnapshot()
  })
})
