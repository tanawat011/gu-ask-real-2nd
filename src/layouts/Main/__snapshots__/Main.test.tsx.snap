// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainLayout /> should render to match snapshot 1`] = `
<DocumentFragment>
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  position: relative;
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgb(17 24 39 / var(--tw-bg-opacity));
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
  min-height: 100vh;
}

.c3 {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgb(31 41 55 / var(--tw-bg-opacity));
}

.c4 {
  position: relative;
  height: 100%;
  overflow-x: auto;
}

<div
    class="c0 c1"
  >
    <div>
      <button
        data-testid="btn2"
      >
        Menu Item
      </button>
    </div>
    <div
      class="c2 c3"
    >
      <div>
        <button
          data-testid="btn1"
        >
          Toggle Sidebar
        </button>
      </div>
      <div
        class="c0 c4"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<MainLayout /> should render to wrong location path 1`] = `
<DocumentFragment>
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  position: relative;
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgb(17 24 39 / var(--tw-bg-opacity));
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
  height: 100vh;
}

.c3 {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgb(31 41 55 / var(--tw-bg-opacity));
}

.c4 {
  position: relative;
  height: 100%;
  overflow-x: auto;
}

<div
    class="c0 c1"
  >
    <div>
      <button
        data-testid="btn2"
      >
        Menu Item
      </button>
    </div>
    <div
      class="c2 c3"
    >
      <div>
        <button
          data-testid="btn1"
        >
          Toggle Sidebar
        </button>
      </div>
      <div
        class="c0 c4"
      />
    </div>
  </div>
</DocumentFragment>
`;
